enterprise/internal/insights/compression/worker.go

```diff
@@ -0,0 +1,183 @@
+package compression
+
+import (
+	"context"
+	"database/sql"
+	"errors"
+	"time"
+
+	"github.com/sourcegraph/sourcegraph/internal/types"
+
+	"github.com/inconshreveable/log15"
+	"github.com/sourcegraph/sourcegraph/cmd/frontend/envvar"
+	"github.com/sourcegraph/sourcegraph/enterprise/internal/insights/discovery"
+	"github.com/sourcegraph/sourcegraph/internal/api"
+	"github.com/sourcegraph/sourcegraph/internal/database"
+	"github.com/sourcegraph/sourcegraph/internal/database/dbcache"
+	"github.com/sourcegraph/sourcegraph/internal/database/dbutil"
+	"github.com/sourcegraph/sourcegraph/internal/goroutine"
+	"github.com/sourcegraph/sourcegraph/internal/vcs/git"
+	"golang.org/x/time/rate"
+)
+
+type RepoStore interface {
+	GetByName(ctx context.Context, repoName api.RepoName) ([]*types.Repo, error)
+}
+
+type CommitIndexer struct {
+	limiter           *rate.Limiter
+	allReposIterator  func(ctx context.Context, each func(repoName string) error) error
+	getRepoId         func(ctx context.Context, name api.RepoName) (*types.Repo, error)
+	getCommits        func(ctx context.Context, name api.RepoName, after time.Time) ([]*git.Commit, error)
+	commitStore       CommitStore
+	maxHistoricalTime time.Time
+	background        context.Context
+}
+
+func NewCommitIndexer(background context.Context, base dbutil.DB, insights dbutil.DB) *CommitIndexer {
+	//todo: add a setting for historical index length
```

Comment:
```suggestion
	// TODO(coury): add a setting for historical index length
```

You will find later on when searching our codebase that you will want a way to find just your TODOs 😆